File: src/main/java/org/jetbrains/fortran/lang/psi/FortranFixedFormFile.java
package org.jetbrains.fortran.lang.psi;

import com.intellij.extapi.psi.PsiFileBase;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.psi.FileViewProvider;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.fortran.FortranFixedFormFileType;
import org.jetbrains.fortran.FortranFixedFormLanguage;

public class FortranFixedFormFile extends PsiFileBase{
    public FortranFixedFormFile(@NotNull FileViewProvider viewProvider) {
        super(viewProvider, FortranFixedFormLanguage.INSTANCE);
    }

    @NotNull
    @Override
    public FileType getFileType() {
        return FortranFixedFormFileType.INSTANCE;
    }
}

File: src/main/java/org/jetbrains/fortran/lang/psi/FortranFixedFormFile.kt
package org.jetbrains.fortran.lang.psi

import com.intellij.extapi.psi.PsiFileBase
import com.intellij.openapi.application.runReadAction
import com.intellij.psi.FileViewProvider
import org.jetbrains.fortran.FortranFixedFormFileType
import org.jetbrains.fortran.FortranFixedFormLanguage
import org.jetbrains.fortran.lang.stubs.FortranProgramUnitStub


class FortranFixedFormFile(viewProvider: FileViewProvider) : PsiFileBase(viewProvider, FortranFixedFormLanguage), FortranCompositeElement {

    override fun getFileType(): FortranFixedFormFileType = FortranFixedFormFileType

    val programUnits : List<FortranProgramUnit>
        get() = runReadAction {
            if (stub != null)
                stub!!.childrenStubs.filter { it is FortranProgramUnitStub }.map { it.psi as FortranProgramUnit }
            else
                children.filter { it is FortranProgramUnit }.map { it as FortranProgramUnit }
        }
}